libs/Matchmaking/src/Peer/PeerDiscoveryAgent.cs

```diff
@@ -588,15 +600,33 @@ internal Client(IPeerDiscoveryTransport net)
             {
                 var handles = new WaitHandle[] { token.WaitHandle, updateAvailable_ };
                 var updatedSubscriptions = new List<DiscoverySubscription>();
+                var currNewHandlers = new List<NewHandler>();
                 while (true)
                 {
                     // Wait for either update or cancellation.
                     WaitHandle.WaitAny(handles);
                     token.ThrowIfCancellationRequested();
 
-                    // There has been an update, collect the dirty tasks.
+                    int currNewHandlersNum = newHandlers_.Count;
+                    currNewHandlers.Capacity = Math.Max(currNewHandlers.Capacity, currNewHandlersNum);
+
                     lock (this)
                     {
+                        // Invoke any newly added handler, even if resources haven't changed.
+                        if (currNewHandlersNum > 0)
```

Comment:
why do we care about currNewHandlersNum etc? vs a dumb TryDequeue loop?

Reply:
A dumb TryDequeue loop would block the thread if a lot of elements would block the thread if a lot of handlers are added at the same time (chances are low, but better to be safe). The `if` here is superflous btw, will remove.

Reply:
OK, so if I understand, we prefer to release the lock, fire some, then go back and drain some more.